### ISSUE TITLE
Implement TrendConstant()

### ISSUE THREAD

**Issue by kennethshsu** [casact/chainladder-python#582]
### Discussed in https://github.com/casact/chainladder-python/discussions/506

<div type='discussions-op-text'>

<sup>Originally posted by **vdunaway** April  4, 2024</sup>
Could someone please provide more details on how the multipart trending works, specifically the math behind what `chainladder` is doing and how to properly enter the dates for what you intend? I've been toggling the start/end dates and it's not working as intended, thus, potentially the documentation/user-guide could be enhanced regarding this.

For example, say I want to apply a multi-step trending that incorporates the following:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/VIRGIN~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/VIRGIN~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#288C6E" vlink="#288C6E">

Trends   for Year X-1 to X* |  
-- | --
  | Annual
Year | Trend Rate
Prior to 2021 | 3.5%
2021 | 10.0%
2023 | 5.0%
  |  
Trend To Date = | 12/31/2024
  |  
*trend   rate remains the same until
new trend in list |  

</body>

</html>

When I run through `chainladder`, I don't get what I was expecting. **Am I inputting the dates incorrectly?**

```
p = pd.DataFrame({'Year': list(range(2014,2024)), 'Premium': [10000]*10})
p_tri = cl.Triangle(p, origin='Year', columns='Premium', cumulative = True)

test = cl.Trend(
    trends=[.05, .1, .035],
    dates=[('2024-12-31', '2022'),('2022', '2020'),('2020','1990')]
).fit(p_tri).trend_

test
Out[71]: 
          2023
2014  1.661630
2015  1.605439
2016  1.551149
2017  1.498695
2018  1.448014
2019  1.399048
2020  1.278334
2021  1.162121
2022  1.102500
2023  1.050000
```

Dissecting the `chainladder` output and comparing to expectation, how would one enter the dates into `cl.Trend()` to obtain the expected result?

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/VIRGIN~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/VIRGIN~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link=blue vlink=purple>

  | Chainladder Output | Chainladder   Resulting YoY Trend Factor | Expected YoY Trend Factor | Expected Result
-- | -- | -- | -- | --
2014 | 1.6616298 | 1.0350000 | 1.0350000 | 1.6398573
2015 | 1.6054394 | 1.0350000 | 1.0350000 | 1.5844032
2016 | 1.5511492 | 1.0350000 | 1.0350000 | 1.5308244
2017 | 1.4986949 | 1.0350000 | 1.0350000 | 1.4790574
2018 | 1.4480144 | 1.0350000 | 1.0350000 | 1.4290409
2019 | 1.3990477 | 1.0944308 | 1.0350000 | 1.3807159
2020 | 1.2783336 | 1.1000000 | 1.1000000 | 1.3340250
2021 | 1.1621214 | 1.0540784 | 1.1000000 | 1.2127500
2022 | 1.1025000 | 1.0500000 | 1.0500000 | 1.1025000
2023 | 1.0500000 | 1.0500000 | 1.0500000 | 1.0500000

</body>

</html>

</div>

**Comment by henrydingliu** [comment]
```
p = pd.DataFrame({'Year': list(range(2014,2025)), 'Premium': [10000]*11})
p_tri = cl.Triangle(p, origin='Year', columns='Premium', cumulative = True)

test = cl.Trend(
    trends=[.05, .1, .035],
    dates=[('2024-12-31', '2022-12-31'),('2022-12-31', '2019-12-31'),('2019-12-31','1990-1-1')]
).fit(p_tri).trend_.latest_diagonal.to_frame()
test['YoY'] = test['2024'] / test.shift(periods=-1)['2024']
test
```


this worked for me

**Comment by henrydingliu** [comment]
I'm looking through Friedland. Chapter 8 Expected Claims and Chapter 11 Freq-Sev both heavily mention applying trend but does not offer any guidance on how to select trend. I would posit that we don't need to implement a trend estimator before 1.0. If we have bandwidth, I actually think implementing apriori estimation into the expected loss IBNR method is a more core need. As a part of that apriori estimation, we'll have a chance to tweak the (end, start) tuple definition to be more aligned with the textbook.

**Comment by kennethshsu** [comment]
A bit more notes on `.TrendConstant()`, it would be nice to have a way to input an array of actual trends, and have that applied to the Triangle object directly, basically, the factors supplied is already the trend, no need to raise to the number of years.

This would be useful when studying tort reform or regulatory impacts, which the text mentions a few times.

**Comment by henrydingliu** [comment]
couldn't you just input an array of actual trends into its own triangle? what would be the advantage of creating a dedicated trendconstant() class if the core use case is just to interact with an actual triangle? the only theoretical scenario would be if you want to somehow involve both the trended and untrended triangle in some estimation. but i've never come across such a need in practice. we are always analyzing just the trended or just the untrended triangle.

**Comment by kennethshsu** [comment]
Ya you could, you basically have a triangle of data and a triangle of the actual trend scalers.

But I thought if you could do something like this it would be easier. This is the example from Friedland page 140 Chapter 8 Exhibit I Sheet I, look at column 11:

<img width="1023" height="366" alt="Image" src="https://github.com/user-attachments/assets/79b6f7ff-379d-428c-8073-694e955d939c" />

```python
cl.Trend(
    trends=[0.670, 0.750, 1.000], 
    dates=[("2003-12-31", "2000-01-01"), ("2004-12-31", "2004-01-01"), ("2008-12-31", "2006-01-01")], 
    annualize=False,
).fit()

```
With a new parameter `annualize` or something like that. And yes, the dates are still weird. What do you think?

**Comment by henrydingliu** [comment]
so try paralellogramolf?

**Comment by kennethshsu** [comment]
Wow you are really smart!

```python
friedland_gl_self_insurer = cl.load_sample("friedland_gl_self_insurer")

tort = pd.DataFrame({
    "EffDate": ["2003-01-01", "2004-01-01", "2005-01-01"],
    "RateChange": [(0.67/0.75) - 1, 0.75-1, 0.00]
})

olf = cl.ParallelogramOLF(
    rate_history=tort,
    change_col="RateChange",
    date_col="EffDate",
    policy_length=12,
    approximation_grain="M",
    vertical_line=True
).fit_transform(friedland_gl_self_insurer["Reported Claims"]).olf_
olf
```
```python
	2008
1998	0.6700
1999	0.6700
2000	0.6700
2001	0.6700
2002	0.6700
2003	0.7500
2004	1.0000
2005	1.0000
2006	1.0000
2007	1.0000
2008	1.0000
```

**Comment by henrydingliu** [comment]
i think we just changed the scope of this issue to "add cumulative to parallelgramolf"

**Comment by kennethshsu** [comment]
We can do that, just opened a new ticket.

I still think we should have rename `.Trend()` to `.TrendConstant()` to match `.Development()` and `.DevelopmentConstant()`. And flip the start and end, that thing is so weird to me.

**Comment by henrydingliu** [comment]
sure. move all of the core codebase to trendconstant, make trend into a one-line redirect, and add a deprecation warning. light work for a first contributor

**Comment by henrydingliu** [comment]
and then once we do that, developmentconstant is gonna look outta place because its the only thing in this package that takes a dict as the main input. maybe we just completely refactor developmentconstant now rather than later

**Comment by henrydingliu** [comment]
@kennethshsu bumping as i came across the `TrendConstant` sketch in the codebase. i feel like we are in a different place than 3 months ago. so let me re-ask the questions to make sure we are still on the right path

- you wanted to make `Trend` into an regression estimator. how does this [example](https://chainladder-python.readthedocs.io/main/friedland/chapter_11.html#exhibit-v-sheet-7) in Chapter 11 meet your requirements?
- you also wanted to migrate the current `Trend` into `TrendConstant`. would you want to rename `Triangle.trend()` as well?

**Comment by kennethshsu** [comment]
I have trend regression in the Bequest Sherman chapter as well, and I'm not very good here, so I don't have an answer to that. But my original idea was that there's a `cl.Trend.fit()` estimator to estimate the trend within the data, we then select a trend, and applying it with `cl.TrendConstant()` to the data. Similar to `cl.Development` and `cl.DevelopmentConstant()`.

`cl.Trend().fit_transform()` could be interesting, that it takes the "default" estimated trend and just apply it to the whole triangle? I'm not sure how this would work in detail, will probably need to see some proposed design to iron it out. 

What do you mean by `Triangle.trend()`?

**Comment by henrydingliu** [comment]
>What do you mean by Triangle.trend()?

It's how you would actually chain trend currently. 

```python
print(raa.trend(0.01))
print(raa.trend(0.01).trend(0.01))
print(raa.trend(0.0201))
```
```python
              12            24            36            48            60            72            84            96            108           120
1981  5481.550587   9043.683520  11928.825269  12910.954644  14807.404907  17696.921397  19696.178076  20351.295554  20410.354559  20598.468426
1982   114.782811   4640.040984   5843.094784  11549.749622  14923.931117  16891.481751  16779.947510  17508.710073  18088.038411           NaN
1983  3655.981551   9640.641086  14873.733740  17305.336718  20086.455822  23816.414712  24512.230555  25158.728173           NaN           NaN
1984  6002.896452  12265.865340  16735.926694  22574.287523  24866.109528  27687.630088  28732.165916           NaN           NaN           NaN
1985  1147.702975  10052.911129  16643.795153  23299.841801  27278.965850  27515.443112           NaN           NaN           NaN           NaN
1986  1574.433867   6706.692844  12177.148125  13460.212869  16495.654767           NaN           NaN           NaN           NaN           NaN
1987   573.877657   4141.810020  11277.674746  12687.126514           NaN           NaN           NaN           NaN           NaN           NaN
1988  1378.155100   7086.634700  13375.551200           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1989  3164.330000   5448.950000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1990  2063.000000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
              12            24            36            48            60            72            84            96            108           120
1981  5995.091148   9890.943476  13046.380517  14120.520950  16194.640673  19354.862304  21541.419890  22257.912228  22322.504191  22528.241557
1982   124.293336   5024.499493   6327.234368  12506.723827  16160.478884  18291.054282  18170.278682  18959.424110  19586.753685           NaN
1983  3919.707067  10336.072125  15946.655760  18553.663276  21535.399385  25534.420173  26280.428937  26973.561888           NaN           NaN
1984  6372.195545  13020.463223  17765.523425  23963.061091  26395.876331  29390.977261  30499.773091           NaN           NaN           NaN
1985  1206.247361  10565.710630  17492.795978  24488.367898  28670.467265  28919.007243           NaN           NaN           NaN           NaN
1986  1638.362196   6979.011468  12671.589169  14006.751487  17165.444498           NaN           NaN           NaN           NaN           NaN
1987   591.266724   4267.311005  11619.399568  13071.559135           NaN           NaN           NaN           NaN           NaN           NaN
1988  1405.856018   7229.076057  13644.399779           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1989  3195.973300   5503.439500           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1990  2063.000000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
              12            24            36            48            60            72            84            96            108           120
1981  5995.091148   9890.943476  13046.380517  14120.520950  16194.640673  19354.862304  21541.419890  22257.912228  22322.504191  22528.241557
1982   124.293336   5024.499493   6327.234368  12506.723827  16160.478884  18291.054282  18170.278682  18959.424110  19586.753685           NaN
1983  3919.707067  10336.072125  15946.655760  18553.663276  21535.399385  25534.420173  26280.428937  26973.561888           NaN           NaN
1984  6372.195545  13020.463223  17765.523425  23963.061091  26395.876331  29390.977261  30499.773091           NaN           NaN           NaN
1985  1206.247361  10565.710630  17492.795978  24488.367898  28670.467265  28919.007243           NaN           NaN           NaN           NaN
1986  1638.362196   6979.011468  12671.589169  14006.751487  17165.444498           NaN           NaN           NaN           NaN           NaN
1987   591.266724   4267.311005  11619.399568  13071.559135           NaN           NaN           NaN           NaN           NaN           NaN
1988  1405.856018   7229.076057  13644.399779           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1989  3195.973300   5503.439500           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
1990  2063.000000           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN           NaN
```

**Comment by kennethshsu** [comment]
Are you asking if we should do it that way, instead of an estimator `cl.TrendConstant()`? Or if you are asking if we would rename the triangle modifier (is this the right word?) to `Triangle.TrendConstant()`? I think `Triangle.Trend()` is self explanatory, `Triangle.TrendConstant()` would be too long.

**Comment by henrydingliu** [comment]
yeah, im asking about renaming `Triangle.trend()`

wouldn't be more confusing when we eventually have `Triangle.trend` and `chainladder.Trend` do different things?

**Comment by kennethshsu** [comment]
Yes, or deprecate the Triangle.Trend? Force them to use the estimator? I have to think about this more...

**Comment by henrydingliu** [comment]
okay. in the meantime, would you mind if we delete trendconstant from main?

**Comment by kennethshsu** [comment]
Ya, I don't know how that even got in there haha, I'll open a PR and move it to a separate branch. I don't even remember what I did so far.
